libsplinter/src/matrix.rs

```diff
@@ -12,54 +12,120 @@
 use std::fmt;
 use std::time::Duration;
 
 use crate::transport::Connection;
 
-/// Wrapper around payload to include connection id
+/// Wrapper around a payload to include connection ID
 #[derive(Debug, Default, PartialEq)]
 pub struct Envelope {
     id: String,
     payload: Vec<u8>,
 }
 
 impl Envelope {
+    /// Create a new Envelope that will be sent over a connection
     pub fn new(id: String, payload: Vec<u8>) -> Self {
         Envelope { id, payload }
     }
 
+    /// Returns the connection ID of the recipient of the envelope
     pub fn id(&self) -> &str {
         &self.id
     }
 
+    /// Return the bytes of the payload
     pub fn payload(&self) -> &[u8] {
         &self.payload
     }
 
+    /// Returns the payload from the message while consuming the the Envelope
     pub fn take_payload(self) -> Vec<u8> {
         self.payload
     }
 }
 
 /// MatrixLifeCycle trait abstracts out adding and removing connections to a
-/// connection handler without requiring knowledge about sending or receiving messges.
+/// connection group without requiring knowledge about sending or receiving messges.
 pub trait MatrixLifeCycle: Clone + Send {
+    /// Add a connection to the connection group with a connection ID
+    ///
+    /// # Arguments
+    ///
+    /// * `connection` - A boxed connection that will be passed to the connection group.
+    /// * `id` - a unique connection ID that will be used to sent messages over a connection.
+    ///
+    /// If the add failed, a MatrixAddError will be returned.
     fn add(&self, connection: Box<dyn Connection>, id: String) -> Result<usize, MatrixAddError>;
+
+    // Removes a connection from the connection group with a connection ID.
+    //
+    /// # Arguments
+    ///
+    /// * `id` - the unique connection ID for the connection that should be removed
+    ///
+    /// If the remove failed, a MatrixRemoveError will be returned.
     fn remove(&self, id: &str) -> Result<Box<dyn Connection>, MatrixRemoveError>;
 }
 
+/// MatrixSender trait abstracts out sending messages through a connection group without
+/// requiring knowledge adding and removing connections, and receiving messages
 pub trait MatrixSender: Clone + Send {
+    /// Send a message over the specified connection.
+    ///
+    /// # Arguments
+    ///
+    /// * `id` - the unique connection ID the message should be sent over
+    /// * `message` - the bytes of the messages
+    ///
+    /// If the send failed, a MatrixSendError will be returned.
     fn send(&self, id: String, message: Vec<u8>) -> Result<(), MatrixSendError>;
 }
 
+/// MatrixReceiver trait abstracts out receiving messages from a connection group without
+/// requiring knowledge adding and removing connections, and sending messages
 pub trait MatrixReceiver: Clone + Send {
+    /// Attempt to receive a message from the connection group. This function will block until
+    /// there is a Envelope to receive. The message will come from the first ready connection
```

Comment:
We should probably put things like "Envolope" between ticks so it shows up as a reference to code in the docs.